media/js/base/send-to-device.js

```diff
@@ -0,0 +1,239 @@
+     */
+    SendToDevice.prototype.enableForm = function() {
+        this.$input.prop('disabled', false);
+        this.$form.removeClass('loading');
+        this.$spinnerTarget.hide();
+    };
+
+    /**
+     * Disable form fields and show loading indicator
+     */
+    SendToDevice.prototype.disableForm = function() {
+        this.$input.prop('disabled', true);
+        this.$form.addClass('loading');
+        this.spinner.spin(this.$spinnerTarget.show()[0]);
+    };
+
+    /**
+     * Reallly primative validation (e.g a@a)
+     * matches built-in validation in Firefox
+     * @param {email}
+     */
+    SendToDevice.prototype.checkEmailValidity = function(email) {
+        return /\S+@\S+/.test(email);
+    };
+
+    /**
+     * Handle form submission via XHR
+     */
+    SendToDevice.prototype.onFormSubmit = function(e) {
+        e.preventDefault();
+
+        var self = this;
+        var action = this.$form.attr('action');
+        var formData = this.$form.serialize();
+
+        this.disableForm();
+
+        // if we know the user has not been prompted to enter an SMS number,
+        // perform some basic email validation before submitting the form.
+        if (!this.smsEnabled && !this.checkEmailValidity(this.$input.val())) {
+            this.onFormError('email');
+            return;
+        }
+
+        // else POST and let the server work out whether the input is a
+        // valid email address or US phone number.
+        $.post(action, formData)
+            .done(function(data) {
+                if (data.success) {
+                    self.onFormSuccess(data.success);
+                } else if (data.errors) {
+                    self.onFormError(data.errors);
+                }
+            })
+            .fail(function(error) {
+                self.onFormFailure(error);
+            });
+    };
+
+    SendToDevice.prototype.onFormSuccess = function() {
+        this.$errorList.addClass('hidden');
+        this.$formFields.addClass('hidden');
+        this.$formHeading.addClass('hidden');
+        this.$thankyou.removeClass('hidden');
+        this.enableForm();
+    };
+
+    SendToDevice.prototype.onFormError = function(errors) {
+        var errorClass;
+        this.$errorList.find('li').hide();
+        this.$errorList.removeClass('hidden');
+
+        if ($.inArray('platform', errors) !== -1) {
+            errorClass = '.platform';
+        } else if (this.smsEnabled && $.inArray('number', errors) !== -1) {
+            errorClass = '.sms';
+        } else {
+            errorClass = '.email';
+        }
+
+        this.$errorList.find(errorClass).show();
+        this.enableForm();
+    };
+
+    SendToDevice.prototype.onFormFailure = function() {
+        this.$errorList.find('li').hide();
+        this.$errorList.removeClass('hidden');
+        this.$errorList.find('.system').show();
+        this.enableForm();
+    };
+
+    window.Mozilla.SendToDevice = SendToDevice;
+
+})(window.jQuery);
```

Comment:
I left this out as it means we don't need to include spinner.js as a dependency in the test files, and can just stub it out instead. I'll add a jshint comment to this file instead to please the linter.